# Artificial Neural Network

### Importing the libraries

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf

In [ ]:
tf.__version__

'2.6.0'

## Part 1 - Data Preprocessing

### Importing the dataset

In [ ]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values #skipping first three columns and last column as it does not affect output result
y = dataset.iloc[:, -1].values

In [ ]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [ ]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2]) #gender column has index 2 now in X

In [ ]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough') #index of geography column is 1
X = np.array(ct.fit_transform(X))

In [ ]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling(it is compulsary for deep learning)

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)
#sc.fit_transform will get mean and standard deviation

## Part 2 - Building the ANN

### Initializing the ANN

In [ ]:
ann = tf.keras.models.Sequential() #Sequential class that allows to build ANN instance(object) as sequential layers using models module in keras library

### Adding the input layer and the first hidden layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) #fully connected layer is the object of Dense class, 6 units is an optimized guess of number of neurons in this first hidden layer
#mostly rectifier activation function(relu) is used in hidden layer for ANN
#Dense class is used to achieve fully connected layers

### Adding the second hidden layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) #1 unit means a binary classification
#mostly sigmoid activation function is used in output layer for ANN(for binary classification)

## Part 3 - Training the ANN

### Compiling the ANN

In [ ]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

### Training the ANN on the Training set

In [ ]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100) #batch size as 32 is classic standard value

Epoch 1/100
250/250 [==============================] - 1s 1ms/step - loss: 0.5943 - accuracy: 0.7237
Epoch 2/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4897 - accuracy: 0.8021
Epoch 3/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4567 - accuracy: 0.8062
Epoch 4/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4353 - accuracy: 0.8146
Epoch 5/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4232 - accuracy: 0.8219
Epoch 6/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4143 - accuracy: 0.8271
Epoch 7/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4077 - accuracy: 0.8266
Epoch 8/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4032 - accuracy: 0.8305
Epoch 9/100
250/250 [==============================] - 0s 1ms/step - loss: 0.3991 - accuracy: 0.8313
Epoch 10/100
250/250 [==============================] - 0s 1ms/step - loss: 0.3948 - accura

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

**Homework**

Use our ANN model to predict if the customer with the following informations will leave the bank: 

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer ?

**Solution**

In [ ]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)
#predict function should always take double square brackets(2d array)
#1, 0, 0 is geography one hot encoding of France
#1 is label encoding of gender male
#sc.transform is used on test dataset as feature scaling was done on training dataset(sc.fit_transform)
#>0.5 is used to get 0 or 1 based on comparison with output of sigmoid function

[[False]]


Therefore, our ANN model predicts that this customer stays in the bank!

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [ ]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))
#1st column=predicted output
#2nd column=actual output

[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1506   89]
 [ 194  211]]


0.8585

1506 - correct predictions that customer stays in bank, 
211 - correct predictions that customer leaves bank, 
89 - incorrect predictions that customer leaves bank, 
194 - incorrect predictions that customer stays in bank